### ISSUE TITLE
Multiple snapshots are incompatible with Raft "catch-up" implementation

### ISSUE THREAD

**Issue by npepinpe** [zeebe-io/zeebe#3573]
**Describe the bug**

The normal Raft algorithm (or at least our implementation) doesn't take into accounts multiple snapshots, so right now we might end up with gaps in our logs, or at least snapshots which should not be recoverable.

Take for example two nodes, L and F, with the following logs:

L: 1, 2, 3, 4, 5, 6, 7, 8, 9, 10
F: 1, 2, 3, 4

L has snapshots S_7, S_8, S_9 (pointing a 7, 8, and 9 respectively), and F has snapshots S_1, S_2, S_3.

There are two cases where L will try to send a snapshot to F:

    F was gone and is catching up
    F is too slow to replicate and keep up

In either case, L will always send its latest snapshot. So when F rejoins (or tries to catch up), it now has snapshots S_1, S_2, S_3, S_9. It will delete S_1 and compact properly, no problem. It will also receive event 10. The problem is that the snapshots S_2 and S_3 are now useless - if we were to recover from them on F (since F can now be leader), we could, but then we'd go from event 4 to event 10 without realizing, resulting in an inconsistent log. This can happen if for whatever reason S_9 is corrupted on F (the whole reason we keep 3 snapshots).

**To Reproduce**

1. Start a cluster of 3 nodes, 1 partition, replication factor 3, max snapshots 3
1. Generate some events, then take a snapshot, and wait for it to be replicated - do this twice (or whatever is max snapshot count - 1)
1. Stop follower node - let's call it F
1. Generate more events, then take a snapshot, and wait for it to be replicated
1. Start F, wait for it to catch up
1. Stop the other follower node
1. Generate more events (this ensures F has a longer log than the other follower)
1. Stop the leader
1. Corrupt the latest snapshot of F (e.g. truncate a file, remove a file, etc.)
1. Start the follower
1. F will now become leader, fail to recover from its latest snapshot and use an older snapshot

This may or may not cause errors (depending on the events generated), but it creates an inconsistent log since some events are missing from the log on replay.

**Expected behavior**

The log is always consistent - we can never replay with gaps in our log; we always recover from a point where we never have gaps after.

**Comment by npepinpe** [comment]
Related to #3531

**Comment by npepinpe** [comment]
@Zelldon 

> Would it help if we send all three snapshots? Or maybe on receiving check whether this snapshot is corrupted, if this is the case we request a new one?

The thing regarding corrupted snapshots is, those can get corrupted due to bit rot. This is very unlikely on persistent disks, but can impact physical disks. But adding a check (opening/closing on receive) is a good idea and we should definitely do it 👍 

Regarding your comment, I think that's the only possible solution - by that I mean, there is no way to keep older snapshots valid once the leader has truncated enough of its log that it simply can't fill the gaps in the follower's log (as far as I can tell - maybe I forgot something 😅). I did some research and checked out other Raft implementations: etcd, JRaft, LogCabin, and NuRaft all only keep one snapshot around, and hashicorp's raft can keep more, but reading their code it looks like they have the same issue as ours (or I couldn't figure out how they manage to handle the potential gaps).

I think with the previous restore procedure we used to just discard the snapshots/log, maybe this is better than nothing 🤷‍♂

**Comment by npepinpe** [comment]
Re: keeping multiple snapshots, the only "correct" solution I can find is that on catch up we replicate all of the leader's snapshots to the follower before considering the instance is truly caught up.

A compromise would be replicating only the oldest snapshot, at which point we can start replicating events from the leader (using the oldest snapshot index to compare as opposed to the current implementation using the latest snapshot index), but we need to discard the follower's older log/snapshots to avoid having inconsistencies. This is what the previous restore procedure was doing, so we could go with this again as a first iteration.

But I'd like first that we really consider whether or not we should keep multiple snapshots around.

**Comment by Zelldon** [comment]
> A compromise would be replicating only the oldest snapshot, at which point we can start replicating events from the leader (using the oldest snapshot index to compare as opposed to the current implementation using the latest snapshot index), but we need to discard the follower's older log/snapshots to avoid having inconsistencies. This is what the previous restore procedure was doing, so we could go with this again as a first iteration.

Sounds good to me. Then after the follower has received newer snapshots (max count) he is able to delete again.

> But I'd like first that we really consider whether or not we should keep multiple snapshots around.

I see no good alternative? Even if we build state on followers. At some point we need to replicate snapshots, because the follower was while gone etc. And to be more fault tolerant regarding to invalid snapshots, we choosed multiple snapshots :man_shrugging: 

Saying that you need to back up your persistent disk periodically sounds like out sourcing the problem to the user. I mean it is possible but this is a product decision, probably we need to discuss this with @menski as well.

**Comment by npepinpe** [comment]
I talked to the cloud guys about this, and taking incremental backups of your persistent disks is probably a no-go. It's a relatively new feature, and it increases the load on the operational side. Personally I still feel like any decent ops team should have some external backup solution, but I guess it's not as standard as I thought and we shouldn't rely on it.

So we have two solutions then:

1. Replicate all snapshots on catch up (e.g. replicate 3 snapshots if max snapshot count is 3) and throw away old state before starting to replicate logs
1. Replicate oldest snapshot only and throw away old state (meaning the follower is not as fault-tolerant as possible until we start receiving more snapshots)

The main difference between both is that for the first, it takes more time to catch up, but once we are caught up the follower is basically a carbon copy of the leader. For the second, once caught up, there is more chance of something going wrong, e.g. the caught up follower becomes leader but the snapshot is corrupted (very unlikely though).

**Comment by npepinpe** [comment]
Latest update:

Current proposed solution is to drop multiple snapshots, and go back to a single snapshot. If, on becoming leader, a node cannot recover from its snapshot, it will step down and throw away its state. By throwing away its state, it cannot become the next leader, and will catch up from the newly elected leader.

This approach offers the same fault tolerance guarantees that Raft normally offers - it's pretty much the same as if that node had suffered full data loss. To mitigate this we should tighten up our snapshot replication to make sure we (nearly) never commit invalid snapshots.

Multiple snapshots in this case only offer slightly more fault tolerance, and really only for the specific case of snapshot corruption - its benefits don't outweigh the additional complexity cost.

**Comment by deepthidevaki** [comment]
> it will step down and throw away its state.

This is to be implemented, right?

> This approach offers the same fault tolerance guarantees that Raft normally offers - it's pretty much the same as if that node had suffered full data loss 

I think so. As long as the other two nodes are already in the cluster and one other node has full data, it can recover.

**Comment by npepinpe** [comment]
@Zelldon opened a PR on the Atomix side to allow a node to step down for a particular partition. Throwing away the state would have to be implemented, but that's pretty simple (should be enough to delete the partition folder, no?). I think it might even be already implemented in `RaftPartition` or `RaftPartitionServer`?

**Comment by deepthidevaki** [comment]
Not sure what are the implications of deleting data while it is running. Delete and restart should work, I guess.

**Comment by npepinpe** [comment]
It's safer, but it's definitely poor UX to have the node self-restart - how would we do this? Delete and exit? :sweat_smile:

**Comment by Zelldon** [comment]
I created a consistency check https://github.com/Zelldon/playground/tree/master/zb-log-check, based on the thing we did with the debugger last week @npepinpe 

I checked a benchmark with max snapshot count 1 and segment size 128 MB with this tool and found that the position drop in between. Not sure if this is a issue with the dispatcher or related to this?


Output:


```
Scan log...
===============
At idx 10108
Current position 180358561660928 (Segment id 41993 segment offset 0)
Is smaller then this last position 218180045888864 (Segment id 50799 segment offset 2219360)
Current lowest 218162869953992 current highest 218180045888864
===============
Scan finished
LOG IS INCONSISTENT!
Lowest position: 180358561660928
Highest position: 218180045888864
Events: 15070
```

**Comment by npepinpe** [comment]
Hm, I hoped that we fixed the dispatcher issues; it could be related to this, depending on what happened to this log, hard to tell without more info. But this is pretty bad :(

**Comment by deepthidevaki** [comment]
> It's safer, but it's definitely poor UX to have the node self-restart - how would we do this? Delete and exit? sweat_smile

:smile: I meant restarting the partition, if it is possible at all.

**Comment by npepinpe** [comment]
Restarting a partition is possible from the Zeebe side - the Atomix partition could be restarted I think but that's not exposed as far as I know.

**Comment by Zelldon** [comment]
I ran last week a benchmark with max snapshot count one and was wondering why it failed and never came back.

![max-one-failed](https://user-images.githubusercontent.com/2758593/72359553-c137d700-36ee-11ea-842b-56f082731a76.png)


The problem was that the brokers on restart where not able to recover from one snapshot, so it seems it is not so unlikely that this happens.


```
I 2020-01-10T10:44:05.551785252Z 2020-01-10 10:44:05.551 [Broker-0-ZeebePartition-2] [Broker-0-zb-actors-2] DEBUG io.zeebe.broker.system - Partition role transitioning from null to FOLLOWER
 
I 2020-01-10T10:44:05.551975107Z 2020-01-10 10:44:05.551 [Broker-0-ZeebePartition-2] [Broker-0-zb-actors-2] DEBUG io.zeebe.broker.system - Installing follower partition service for partition PartitionId{id=2, group=raft-partition}
 
I 2020-01-10T10:44:05.567702146Z 2020-01-10 10:44:05.567 [Broker-0-ZeebePartition-2] [Broker-0-zb-actors-2] DEBUG io.zeebe.logstreams.snapshot - Available snapshots: [SnapshotImpl{position=208701057796576, path=/usr/local/zeebe/data/raft-partition/partitions/2/snapshots/87676202-1-1578652300194-208701057796576}]
 
I 2020-01-10T10:44:05.856577870Z 2020-01-10 10:44:05.840 [Broker-0-ZeebePartition-2] [Broker-0-zb-actors-2] ERROR io.zeebe.logstreams.snapshot - Failed to open snapshot 'SnapshotImpl{position=208701057796576, path=/usr/local/zeebe/data/raft-partition/partitions/2/snapshots/87676202-1-1578652300194-208701057796576}'. No snapshots available to recover from. Manual action is required.
 
I 2020-01-10T10:44:05.856607030Z java.lang.RuntimeException: Unexpected error occurred trying to open the database
	at io.zeebe.db.impl.rocksdb.ZeebeRocksDbFactory.open(ZeebeRocksDbFactory.java:82) ~[zeebe-db-0.22.0-SNAPSHOT.jar:0.22.0-SNAPSHOT]
	at io.zeebe.db.impl.rocksdb.ZeebeRocksDbFactory.createDb(ZeebeRocksDbFactory.java:45) ~[zeebe-db-0.22.0-SNAPSHOT.jar:0.22.0-SNAPSHOT]
	at io.zeebe.db.impl.rocksdb.ZeebeRocksDbFactory.createDb(ZeebeRocksDbFactory.java:24) ~[zeebe-db-0.22.0-SNAPSHOT.jar:0.22.0-SNAPSHOT]
	at io.zeebe.logstreams.state.StateSnapshotController.openDb(StateSnapshotController.java:149) ~[zeebe-logstreams-0.22.0-SNAPSHOT.jar:0.22.0-SNAPSHOT]
	at io.zeebe.logstreams.state.StateSnapshotController.recover(StateSnapshotController.java:120) ~[zeebe-logstreams-0.22.0-SNAPSHOT.jar:0.22.0-SNAPSHOT]
	at io.zeebe.broker.system.partitions.ZeebePartition.basePartitionInstallation(ZeebePartition.java:267) ~[zeebe-broker-0.22.0-SNAPSHOT.jar:0.22.0-SNAPSHOT]
	at io.zeebe.broker.system.partitions.ZeebePartition.installFollowerPartition(ZeebePartition.java:214) ~[zeebe-broker-0.22.0-SNAPSHOT.jar:0.22.0-SNAPSHOT]
	at io.zeebe.broker.system.partitions.ZeebePartition.lambda$transitionToFollower$6(ZeebePartition.java:206) ~[zeebe-broker-0.22.0-SNAPSHOT.jar:0.22.0-SNAPSHOT]
	at io.zeebe.util.sched.future.FutureContinuationRunnable.run(FutureContinuationRunnable.java:28) [zeebe-util-0.22.0-SNAPSHOT.jar:0.22.0-SNAPSHOT]
	at io.zeebe.util.sched.ActorJob.invoke(ActorJob.java:76) [zeebe-util-0.22.0-SNAPSHOT.jar:0.22.0-SNAPSHOT]
	at io.zeebe.util.sched.ActorJob.execute(ActorJob.java:39) [zeebe-util-0.22.0-SNAPSHOT.jar:0.22.0-SNAPSHOT]
	at io.zeebe.util.sched.ActorTask.execute(ActorTask.java:115) [zeebe-util-0.22.0-SNAPSHOT.jar:0.22.0-SNAPSHOT]
	at io.zeebe.util.sched.ActorThread.executeCurrentTask(ActorThread.java:107) [zeebe-util-0.22.0-SNAPSHOT.jar:0.22.0-SNAPSHOT]
	at io.zeebe.util.sched.ActorThread.doWork(ActorThread.java:91) [zeebe-util-0.22.0-SNAPSHOT.jar:0.22.0-SNAPSHOT]
	at io.zeebe.util.sched.ActorThread.run(ActorThread.java:195) [zeebe-util-0.22.0-SNAPSHOT.jar:0.22.0-SNAPSHOT]
Caused by: org.rocksdb.RocksDBException: While opening a file for sequentially reading: /usr/local/zeebe/data/raft-partition/partitions/2/runtime/MANIFEST-007879: No such file or directory
	at org.rocksdb.OptimisticTransactionDB.open(Native Method) ~[rocksdbjni-6.5.2.jar:?]
	at org.rocksdb.OptimisticTransactionDB.open(OptimisticTransactionDB.java:80) ~[rocksdbjni-6.5.2.jar:?]
	at io.zeebe.db.impl.rocksdb.transaction.ZeebeTransactionDb.openTransactionalDb(ZeebeTransactionDb.java:87) ~[zeebe-db-0.22.0-SNAPSHOT.jar:0.22.0-SNAPSHOT]
	at io.zeebe.db.impl.rocksdb.ZeebeRocksDbFactory.open(ZeebeRocksDbFactory.java:74) ~[zeebe-db-0.22.0-SNAPSHOT.jar:0.22.0-SNAPSHOT]
	... 14 more
 
I 2020-01-10T10:44:05.857443720Z 2020-01-10 10:44:05.856 [Broker-0-ZeebePartition-2] [Broker-0-zb-actors-2] DEBUG io.zeebe.util.actor - Continuing on future completion failed
 
I 2020-01-10T10:44:05.857467389Z java.lang.IllegalStateException: Unexpected error occurred while recovering snapshot controller during leader partition install for partition 2
	at io.zeebe.broker.system.partitions.ZeebePartition.basePartitionInstallation(ZeebePartition.java:272) ~[zeebe-broker-0.22.0-SNAPSHOT.jar:0.22.0-SNAPSHOT]
	at io.zeebe.broker.system.partitions.ZeebePartition.installFollowerPartition(ZeebePartition.java:214) ~[zeebe-broker-0.22.0-SNAPSHOT.jar:0.22.0-SNAPSHOT]
	at io.zeebe.broker.system.partitions.ZeebePartition.lambda$transitionToFollower$6(ZeebePartition.java:206) ~[zeebe-broker-0.22.0-SNAPSHOT.jar:0.22.0-SNAPSHOT]
	at io.zeebe.util.sched.future.FutureContinuationRunnable.run(FutureContinuationRunnable.java:28) [zeebe-util-0.22.0-SNAPSHOT.jar:0.22.0-SNAPSHOT]
	at io.zeebe.util.sched.ActorJob.invoke(ActorJob.java:76) [zeebe-util-0.22.0-SNAPSHOT.jar:0.22.0-SNAPSHOT]
	at io.zeebe.util.sched.ActorJob.execute(ActorJob.java:39) [zeebe-util-0.22.0-SNAPSHOT.jar:0.22.0-SNAPSHOT]
	at io.zeebe.util.sched.ActorTask.execute(ActorTask.java:115) [zeebe-util-0.22.0-SNAPSHOT.jar:0.22.0-SNAPSHOT]
	at io.zeebe.util.sched.ActorThread.executeCurrentTask(ActorThread.java:107) [zeebe-util-0.22.0-SNAPSHOT.jar:0.22.0-SNAPSHOT]
	at io.zeebe.util.sched.ActorThread.doWork(ActorThread.java:91) [zeebe-util-0.22.0-SNAPSHOT.jar:0.22.0-SNAPSHOT]
	at io.zeebe.util.sched.ActorThread.run(ActorThread.java:195) [zeebe-util-0.22.0-SNAPSHOT.jar:0.22.0-SNAPSHOT]
Caused by: java.lang.RuntimeException: Failed to recover from snapshots
	at io.zeebe.logstreams.state.StateSnapshotController.recover(StateSnapshotController.java:137) ~[zeebe-logstreams-0.22.0-SNAPSHOT.jar:0.22.0-SNAPSHOT]
	at io.zeebe.broker.system.partitions.ZeebePartition.basePartitionInstallation(ZeebePartition.java:267) ~[zeebe-broker-0.22.0-SNAPSHOT.jar:0.22.0-SNAPSHOT]
	... 9 more
Caused by: java.lang.RuntimeException: Unexpected error occurred trying to open the database
	at io.zeebe.db.impl.rocksdb.ZeebeRocksDbFactory.open(ZeebeRocksDbFactory.java:82) ~[zeebe-db-0.22.0-SNAPSHOT.jar:0.22.0-SNAPSHOT]
	at io.zeebe.db.impl.rocksdb.ZeebeRocksDbFactory.createDb(ZeebeRocksDbFactory.java:45) ~[zeebe-db-0.22.0-SNAPSHOT.jar:0.22.0-SNAPSHOT]
	at io.zeebe.db.impl.rocksdb.ZeebeRocksDbFactory.createDb(ZeebeRocksDbFactory.java:24) ~[zeebe-db-0.22.0-SNAPSHOT.jar:0.22.0-SNAPSHOT]
	at io.zeebe.logstreams.state.StateSnapshotController.openDb(StateSnapshotController.java:149) ~[zeebe-logstreams-0.22.0-SNAPSHOT.jar:0.22.0-SNAPSHOT]
	at io.zeebe.logstreams.state.StateSnapshotController.recover(StateSnapshotController.java:120) ~[zeebe-logstreams-0.22.0-SNAPSHOT.jar:0.22.0-SNAPSHOT]
	at io.zeebe.broker.system.partitions.ZeebePartition.basePartitionInstallation(ZeebePartition.java:267) ~[zeebe-broker-0.22.0-SNAPSHOT.jar:0.22.0-SNAPSHOT]
	... 9 more
Caused by: org.rocksdb.RocksDBException: While opening a file for sequentially reading: /usr/local/zeebe/data/raft-partition/partitions/2/runtime/MANIFEST-007879: No such file or directory
	at org.rocksdb.OptimisticTransactionDB.open(Native Method) ~[rocksdbjni-6.5.2.jar:?]
	at org.rocksdb.OptimisticTransactionDB.open(OptimisticTransactionDB.java:80) ~[rocksdbjni-6.5.2.jar:?]
	at io.zeebe.db.impl.rocksdb.transaction.ZeebeTransactionDb.openTransactionalDb(ZeebeTransactionDb.java:87) ~[zeebe-db-0.22.0-SNAPSHOT.jar:0.22.0-SNAPSHOT]
	at io.zeebe.db.impl.rocksdb.ZeebeRocksDbFactory.open(ZeebeRocksDbFactory.java:74) ~[zeebe-db-0.22.0-SNAPSHOT.jar:0.22.0-SNAPSHOT]
	at io.zeebe.db.impl.rocksdb.ZeebeRocksDbFactory.createDb(ZeebeRocksDbFactory.java:45) ~[zeebe-db-0.22.0-SNAPSHOT.jar:0.22.0-SNAPSHOT]
	at io.zeebe.db.impl.rocksdb.ZeebeRocksDbFactory.createDb(ZeebeRocksDbFactory.java:24) ~[zeebe-db-0.22.0-SNAPSHOT.jar:0.22.0-SNAPSHOT]
	at io.zeebe.logstreams.state.StateSnapshotController.openDb(StateSnapshotController.java:149) ~[zeebe-logstreams-0.22.0-SNAPSHOT.jar:0.22.0-SNAPSHOT]
	at io.zeebe.logstreams.state.StateSnapshotController.recover(StateSnapshotController.java:120) ~[zeebe-logstreams-0.22.0-SNAPSHOT.jar:0.22.0-SNAPSHOT]
	at io.zeebe.broker.system.partitions.ZeebePartition.basePartitionInstallation(ZeebePartition.java:267) ~[zeebe-broker-0.22.0-SNAPSHOT.jar:0.22.0-SNAPSHOT]
	... 9 more
 
I 2020-01-10T10:44:07.086892097Z 2020-01-10 10:44:07.086 [exporter-3] [Broker-0-zb-fs-workers-0] DEBUG io.zeebe.broker.exporter - Recovered exporter 'exporter-3' from snapshot at lastExportedPosition 139745354914520
 
I 2020-01-10T10:44:07.092464926Z 2020-01-10 10:44:07.092 [exporter-3] [Broker-0-zb-fs-workers-0] DEBUG io.zeebe.broker.exporter - Set event filter for exporters: ExporterEventFilter{acceptRecordTypes={COMMAND_REJECTION=false, COMMAND=false, EVENT=true, NULL_VAL=false, SBE_UNKNOWN=false}, acceptValueTypes={VARIABLE=true, VARIABLE_DOCUMENT=true, WORKFLOW_INSTANCE_SUBSCRIPTION=false, MESSAGE_SUBSCRIPTION=false, SBE_UNKNOWN=false, JOB_BATCH=false, ERROR=true, WORKFLOW_INSTANCE_RESULT=false, INCIDENT=true, MESSAGE=false, DEPLOYMENT=true, WORKFLOW_INSTANCE_CREATION=false, TIMER=false, JOB=true, MESSAGE_START_EVENT_SUBSCRIPTION=false, WORKFLOW_INSTANCE=true, NULL_VAL=false}}
 
I 2020-01-10T10:44:08.409478092Z 2020-01-10 10:44:08.409 [partition-3-processor] [Broker-0-zb-actors-0] INFO  io.zeebe.processor - Processor starts reprocessing, until last source event position 140024524592024
 
I 2020-01-10T10:44:33.659618447Z 2020-01-10 10:44:33.658 [] [main] INFO  io.zeebe.broker.system - Bootstrap Broker-0 partitions [2/3]: partition 2 failed with unexpected exception.
 
I 2020-01-10T10:44:33.660151686Z 2020-01-10 10:44:33.659 [] [main] INFO  io.zeebe.broker.system - Closing Broker-0 partitions [1/1]: partition 3
 
I 2020-01-10T10:44:33.661192711Z 2020-01-10 10:44:33.660 [Broker-0-ZeebePartition-3] [Broker-0-zb-actors-3] DEBUG io.zeebe.broker.system - Closing exporter-3
 
I 2020-01-10T10:44:33.752368248Z 2020-01-10 10:44:33.752 [exporter-3] [Broker-0-zb-fs-workers-2] DEBUG io.zeebe.broker.exporter - Closed exporter director 'exporter-3'.
 
```

The brokers are then endless restarted, because they were not able to install one of 3 partitions. Might be sufficient to step down at that case.

**Comment by npepinpe** [comment]
Closing this for now as it is being phased out
